docs/gguf.md

```diff
@@ -0,0 +1,546 @@
+
+- `general.source.url: string`: URL to the source of the model. Can be a GitHub repo, a paper, etc.
+- `general.source.huggingface.repository: string`: Hugging Face model repository that this model is either hosted on or based on
+
+### LLM
+
+In the following, `[llm]` is used to fill in for the name of a specific LLM architecture. They will be used in each architecture's section.
+
+- `[llm].context_length: uint32`: Also known as `n_ctx`. length of the context (in tokens) that the model was trained on. For most architectures, this is the hard limit on the length of the input. Architectures, like RWKV, that are not reliant on transformer-style attention may be able to handle larger inputs, but this is not guaranteed.
+- `[llm].embedding_length: uint32`: Also known as `n_embd`. Embedding layer size.
+- `[llm].block_count: uint32`: The number of blocks of attention+feedforward layers (i.e. the bulk of the LLM). Does not include the input or embedding layers.
+- `[llm].feedforward_length: uint32`: Also known as `n_ff`. The length of the feedforward layer.
+- `[llm].use_parallel_residual: bool`: Whether or not the parallel residual logic should be used.
+- `[llm].tensor_data_layout: string`: When a model is converted to GGUF, tensors may be rearranged to improve performance. This key describes the layout of the tensor data. This is not required; if not present, it is assumed to be `reference`.
+  - `reference`: tensors are laid out in the same order as the original model
+  - further options can be found for each architecture in their respective sections
+
+#### Attention
+
+- `[llm].attention.head_count: uint32`: Also known as `n_head`. Number of attention heads.
+- `[llm].attention.head_count_kv: uint32`: The number of heads per group used in Grouped-Query-Attention. If not present, the model does not use GQA.
+- `[llm].attention.max_alibi_bias: float32`: The maximum bias to use for ALiBI.
+- `[llm].attention.clamp_kqv: float32`: Value (`C`) to clamp the values of the `Q`, `K`, and `V` tensors between (`[-C, C]`).
+- `[llm].attention.layer_norm_epsilon: float32`: Layer normalization epsilon.
+- `[llm].attention.layer_norm_rms_epsilon: float32`: Layer RMS normalization epsilon.
+
+#### RoPE
+
+- `[llm].rope.dimension_count: uint32`: The number of rotary dimensions for RoPE.
+- `[llm].rope.scale: float32`: A scale factor for RoPE to adjust the context length.
+
+#### Models
+
+The following sections describe the metadata for each model architecture. Each key specified _must_ be present.
+
+##### LLaMA
+
+- `llama.context_length`
+- `llama.embedding_length`
+- `llama.layer_count`
+- `llama.feedforward_length`
+- `llama.rope.dimension_count`
+- `llama.attention.head_count`
+- `llama.attention.layer_norm_rms_epsilon`
+
+###### Optional
+
+- `llama.rope.scale`
+- `llama.attention.head_count_kv`
+- `llama.tensor_data_layout`:
+  - `llama.cpp`:
```

Comment:
The format in use is actually the format of the original .pth LLaMA models from Meta AI. The permute transforms HF models back to the original pth format.

Reply:
This is the only one I'm not sure about - what's `llama.cpp` producing for this metadata value right now?

Reply:
[pth --> gguf script](https://github.com/ggerganov/llama.cpp/blob/gguf/convert-llama-7b-pth-to-gguf.py)
[hf --> gguf script](https://github.com/ggerganov/llama.cpp/blob/gguf/convert-llama-h5-to-gguf.py)

Reply:
In https://github.com/ggerganov/llama.cpp/pull/2398 the value for this is now `Meta AI original pth`